js/util/data-fetcher.js

```diff
@@ -0,0 +1,111 @@
+ * A copy of the GNU Lesser General Public License is in the file COPYING.
+ */
+
+/** @ignore */
+var Interest = require('../interest.js').Interest; /** @ignore */
+var LOG = require('../log.js').Log.LOG;
+
+/**
+ * DataFetcher is a utility class to fetch Data with automatic retries.
+ *
+ * This is a public constructor to create a new DataFetcher object.
+ * @param {Face} face The segment will be fetched through this face.
+ * @param {Interest} interest Use this as the basis of the future issued Interest(s) to fetch
+ * the solicited segment.
+ * @param {int} maxTimeoutRetries The max number of retries upon facing Timeout.
+ * @param {int} maxNackRetries The max number of retries upon facing Nack.
+ * @param {function} onData Call this function upon receiving the Data packet for the
+ * solicited segment.
+ * @param {function} onTimeout Call this function after receiving Timeout for more than
+ * maxTimeoutRetries times. 
+ * @param {function} onNack Call this function after receiving Nack for more than
+ * maxNackRetries times.
+ * @constructor
+ *
+ */
+var DataFetcher = function DataFetcher
+  (face, interest, maxTimeoutRetries, maxNackRetries, onData, onTimeout, onNack)
+{
+  this.face = face;
+  this.interest = interest;
+  this.maxNackRetries = maxNackRetries;
+  this.maxTimeoutRetries = maxTimeoutRetries;
+
+  this.onData = onData;
+  this.onTimeout = onTimeout;
+  this.onNack = onNack;
+
+  this.numberOfTimeoutRetries = 0;
+  this.pendingInterestId = null;
+};
+
+exports.DataFetcher = DataFetcher;
+
+DataFetcher.prototype.fetch = function()
+{
+  this.pendingInterestId = this.face.expressInterest
+    (this.interest,
+     this.handleData.bind(this),
+     this.handleTimeout.bind(this),
+     this.handleNack.bind(this));
+};
+
+DataFetcher.prototype.cancelPendingInterest = function()
+{
+  this.face.removePendingInterest(this.pendingInterestId);
+};
+
+DataFetcher.prototype.handleData = function(originalInterest, data)
+{
+  this.onData(originalInterest, data);
+};
+
+DataFetcher.prototype.handleTimeout = function(interest)
+{
+  this.numberOfTimeoutRetries++;
+  if(this.numberOfTimeoutRetries <= this.maxTimeoutRetries) {
+    var newInterest = new Interest(interest);
+    // Changing a field clears the nonce so that a new nonce will be generated
+    newInterest.setMustBeFresh(false);
+    this.interest = newInterest;
+    if (LOG > 3)
+      console.log('handle timeout for interest ' + interest.getName());
+    this.fetch();
+  }
+  else {
+    this.onTimeout(interest);
+  }
+};
+
+DataFetcher.prototype.handleNack = function(interest)
+{
+  this.numberOfNackRetries += 1;
+  if(this.numberOfNackRetries <= this.maxNackRetries) {
+    var newInterest = new Interest(interest);
+    // Changing a field clears the nonce so that a new nonce will be generated
+    newInterest.setMustBeFresh(false);
+    this.interest = newInterest;
+    if (LOG > 3)
+      console.log('handle nack for interest ' + interest.getName());
+    this.fetch();
```

Comment:
It's a very bad idea to retransmit right away upon Nack, see https://redmine.named-data.net/issues/4712#note-1
At least add a randomized delay (`setTimeout(this.fetch.bind(this), 40 + Math.random() * 20)` gives you a delay between 40 and 60 milliseconds); exponential back-off is even better.